src/rgw/rgw_sync.cc

```diff
@@ -1498,68 +1430,122 @@ class RGWMetaSyncShardControlCR : public RGWBackoffControlCR
+                const rgw_meta_sync_status& _sync_status)
+    : RGWCoroutine(_sync_env->cct), sync_env(_sync_env),
+      pool(sync_env->store->get_zone_params().log_pool),
+      cursor(cursor), sync_status(_sync_status) {}
 
   int operate() {
+    int ret = 0;
     reenter(this) {
-      yield {
-	map<uint32_t, rgw_meta_sync_marker>::iterator iter = sync_status.sync_markers.begin();
-	for (; iter != sync_status.sync_markers.end(); ++iter) {
-	  uint32_t shard_id = iter->first;
-	  rgw_meta_sync_marker marker;
-
-	  RGWMetaSyncShardControlCR *shard_cr = new RGWMetaSyncShardControlCR(sync_env,
-                                                                              sync_env->store->get_zone_params().log_pool,
-                                                                              shard_id,
-                                                                              sync_status.sync_markers[shard_id]);
+      // loop through one period at a time
+      for (;;) {
+        if (cursor == sync_env->store->period_history->get_current()) {
+          next = RGWPeriodHistory::Cursor{};
+          if (cursor) {
+            ldout(cct, 10) << "RGWMetaSyncCR on current period="
+                << cursor.get_period().get_id() << dendl;
+          } else {
+            ldout(cct, 10) << "RGWMetaSyncCR with no period" << dendl;
+          }
+        } else {
+          next = cursor;
+          next.next();
+          ldout(cct, 10) << "RGWMetaSyncCR on period="
+              << cursor.get_period().get_id() << ", next="
+              << next.get_period().get_id() << dendl;
+        }
 
+        yield {
+          // get the mdlog for the current period (may be empty)
+          auto& period_id = sync_status.sync_info.period;
+          auto mdlog = sync_env->store->meta_mgr->get_log(period_id);
+
+          // sync this period on each shard
+          for (const auto& m : sync_status.sync_markers) {
+            uint32_t shard_id = m.first;
+            auto& marker = m.second;
+
+            std::string period_marker;
+            if (next) {
+              // read the maximum marker from the next period's sync status
+              period_marker = next.get_period().get_sync_status()[shard_id];
+              if (period_marker.empty()) {
+                // no metadata changes have occurred on this shard, skip it
+                ldout(cct, 10) << "RGWMetaSyncCR: skipping shard " << shard_id
+                    << " with empty period marker" << dendl;
+                continue;
+              }
+            }
 
-	  shard_crs[shard_id] = shard_cr;
-          spawn(shard_cr, true);
+            auto cr = new RGWMetaSyncShardControlCR(sync_env, pool, period_id,
+                                                    mdlog, shard_id, marker,
+                                                    std::move(period_marker));
+            // XXX: do we need to hold a ref on cr while it's in shard_crs?
```

Comment:
@cbodley originally we didn't need it, because we spawned with wait=true. Now I'm not 100% sure. Can probably avoid it, as long as we drain all before finishing.